### PR TITLE
Apm stuff

### PR DESCRIPTION
**Pull request by tcanabrava**
This is a work in progress dialog to set the correct airframe type on an APM device.
I based this code on the APM Planner airframe configuration but code is 98% mine.
Caveats:
 - slow, a few requests to ardupilot git repository to get the parameters file, probably I should cache them
 - when you click on Airframes config, screen will take around 10s to display because of that.
 - when you click on 'Apply and reestart', 'Yes', the parameters will appear on the debug output and not apply. I was not confident that the code to apply was correct so I commented it out.
 - I had to create a QAbstractListModel and remove the QVariantList for the AirframeTypes because they are created after the download finishes.
 - XML file for the APM drones are handmade and probably with the wrong configuration for the Type (just need a finetuning )

### CODE REVIEW

File: src/AutoPilotPlugins/APM/APMAirframeComponentController.h [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    QString name() const;
+    QString imageResource() const;
+
+private:
+    QString         _name;
+    QString         _imageResource;
+    QVariantList    _airframes;
+};
+
+class APMAirframeModel : public QAbstractListModel {
```
Comment (DonLakeFlyer) [inline src/AutoPilotPlugins/APM/APMAirframeComponentController.h]: You don't need to build one of these yourself. QmlObjectListModel.cc/h alraedy exists which supports creating a QmlObjectListModel from any QObject.

**Comment by DonLakeFlyer** [comment]
I don't think the APM stack needs a reboot after you apply. Mission Planner doesn't do that does it?

**Comment by DonLakeFlyer** [comment]
How does MissionPlanner or APMPlanner deal with the param file download thing? Does it download it every single time or does it do some sort of check. You could ask Bill on bitter how APM Planner handles it. In general this should be handled like param meta data. There is a version compiled into QGC. And then newer versions are downloaded on the fly and cached.

File: src/AutoPilotPlugins/APM/AirframeFactMetaData.xml [hunk 2]
```diff
@@ -0,0 +1,70 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<airframes>
+  <version>1</version>
+  <airframe_group image="AirframeQuadRotorX.png" name="Quadrotor x">
+      <airframe id="10020" maintainer="Lorenz Meier &lt;lorenz@px4.io&gt;" name="3DR Aero M" file="3DR_AERO_M.param">
```
Comment (DonLakeFlyer) [inline src/AutoPilotPlugins/APM/AirframeFactMetaData.xml]: Is the maintainer field required? I'd just leave it out.

File: src/AutoPilotPlugins/APM/APMAirframeComponentController.h (hunk 1 above)
Comment (tcanabrava) [inline src/AutoPilotPlugins/APM/APMAirframeComponentController.h]: I discovered that when I had already finished this one, I'll update the code accordingly.

File: src/AutoPilotPlugins/APM/AirframeFactMetaData.xml (hunk 2 above)
Comment (tcanabrava) [inline src/AutoPilotPlugins/APM/AirframeFactMetaData.xml]: okay.

**Comment by tcanabrava** [comment]
@DonLakeFlyer it doesn't needs a reboot, I'm removing the check for the multiple devices. about the parameters file, it downloads it everytime.

**Comment by LorenzMeier** [comment]
Can you please add a few screenshots of the thing in operation? We commonly not only review the code, but also the visual quality and usability and it would be great to understand if we're already there with this. Thanks!

**Comment by tcanabrava** [comment]
The visual is the same as the px4 plugin, I'll add one on the new pr that
I'm working on
Em 18 de nov de 2015 06:40, "Lorenz Meier" <notifications@github.com>
escreveu:

**Comment by DonLakeFlyer** [comment]
Let me know when you have the new changes. Then I'll build it myself and take a deeper look.

**Comment by tcanabrava** [comment]
closed because of a new version

**Pull request closed without merging**